language/evm/move-to-yul/src/events.rs

```diff
@@ -186,6 +186,168 @@ impl EventSignature {
     }
 }
 
+/// Generate event emit functions for send_<name> in async move contracts
+/// example: the function send_foo(actor: address, args) will emit an ethereum event
+/// Foo(actor, message_hash, args) where encoding follows the ethereum standard
+pub(crate) fn define_emit_fun_for_send(
+    gen: &mut FunctionGenerator,
+    ctx: &Context,
+    event_sig: &EventSignature,
+    fun_id: &QualifiedInstId<FunId>,
+) {
+    let fun = ctx.env.get_function(fun_id.to_qualified_id());
+    let target = &ctx.targets.get_target(&fun, &FunctionVariant::Baseline);
+
+    // Emit function header
+    let params = (0..target.get_parameter_count()).map(|idx| ctx.make_local_name(target, idx));
+    let params_str = params.clone().join(",");
+    let params_vec = params.collect_vec();
+
+    let mut local_name_idx = target.get_parameter_count();
+
+    emit!(
+        ctx.writer,
+        "function {}({}) ",
+        ctx.make_function_name(fun_id),
+        params_str
+    );
+
+    // Generate the function body
+    ctx.emit_block(|| {
+        let signature_types = &event_sig.para_types;
+        let topic_0_var = ctx.make_local_name(target, local_name_idx);
+        local_name_idx += 1;
+        let event_sig_str = event_sig.to_string();
+        let topic_0_hash = format!("0x{:x}", Keccak256::digest(event_sig_str.as_bytes()));
+
+        emitln!(ctx.writer, "let {} := {}", topic_0_var, topic_0_hash);
+
+        // Compute the message hash following the logic in move_compiler::attr_derivation::async_deriver::message_hash
+        // message_hash is the first 8 bytes of Sha3_256::digest(address::module_name::foo)
+        // example: 0x00000000000000000000000000000003::AccountStateMachine::deposit
+        let message_str = format!(
+            "0x{}::{}",
+            fun.module_env.self_address(),
+            fun.get_full_name_str().replace("send_", "")
+        );
+        let hash_bytes = Sha3_256::digest(message_str.as_bytes());
+        let message_hash_str = &format!("0x{:x}", hash_bytes)[0..10];
+        let message_hash_var = ctx.make_local_name(target, local_name_idx);
+        local_name_idx += 1;
+        emitln!(
+            ctx.writer,
+            "let {} := {}",
+            message_hash_var,
+            message_hash_str
+        );
+
+        let mut indexed_paras = vec![];
+        let mut indexed_vars = vec![topic_0_var];
+        let mut unindexed_paras = vec![];
+        let mut unindexed_vars = vec![];
+
+        for (i, (_, solidity_ty, move_ty, indexed_flag, _)) in signature_types.iter().enumerate() {
+            let mut var = if i == 0 {
+                params_vec.get(0).unwrap().to_string()
+            } else if i == 1 {
+                message_hash_var.clone()
+            } else {
+                params_vec.get(i - 1).unwrap().to_string()
+            };
+
+            if *indexed_flag {
```

Comment:
Can you give me an example when this could be indexed? IIUC we don't need to index any of the extra arguments for param.

Reply:
I am not sure what you mean here but we can add indexed to parameters if there is a need.

Reply:
IIUC the arguments here would be receipient address, function hash, and the rest of the arguments. Are we going to index any of those parameters? Can module author specific which arguments to index?